src/Data/Argonaut/Decode/Implementation.purs

```diff
@@ -0,0 +1,163 @@
+module Data.Argonaut.Decode.Implementation where
+
+import Prelude
+
+import Data.Argonaut.Core (Json, caseJsonBoolean, caseJsonNull, caseJsonNumber, caseJsonString, isNull, stringify, toArray, toObject, toString)
+import Data.Array as Arr
+import Data.Array.NonEmpty (NonEmptyArray)
+import Data.Array.NonEmpty as NEA
+import Data.Bifunctor (lmap, rmap)
+import Data.Either (Either(..), note)
+import Data.Identity (Identity(..))
+import Data.Int (fromNumber)
+import Data.List (List, fromFoldable)
+import Data.List as L
+import Data.List.NonEmpty (NonEmptyList)
+import Data.List.NonEmpty as NEL
+import Data.Map as M
+import Data.Maybe (maybe, Maybe(..))
+import Data.NonEmpty (NonEmpty, (:|))
+import Data.Set as S
+import Data.String (CodePoint, codePointAt)
+import Data.Traversable (traverse)
+import Data.TraversableWithIndex (traverseWithIndex)
+import Data.Tuple (Tuple(..))
+import Foreign.Object as FO
+
+type Decoder a = Json -> Either String a
```

Comment:
I'm on the fence about introducing a type synonym for `Json -> Either String a`; I personally prefer to introduce type synonyms only when they significantly aid readability. It's not clear seeing `Decoder a` that it's a function, and so I would need to look up the type synonym to see what it is. I don't think the readability argument is that strong in this case.

That said, I know that my preference isn't necessarily shared. [Elm uses the `Decoder` type](https://guide.elm-lang.org/effects/json.html), for example.

Reply:
DecoderFunction? It prettifies implementation module a lot

Reply:
If we use type synonyms then `Encoder` and `Decoder` are good names. I only commented because we haven't used synonyms before, using just the actual type `Json -> Either String a`, and in I often prefer seeing the actual type to a synonym.

That said, these synonyms are pretty clear and they do make things prettier.

Reply:
FWIW I strongly agree with @thomashoneyman. In my experience, obscuring a function signature via a type synonym does more harm than good. `Json -> Either String a` is hardly more verbose than `Decoder a`, and it's 100% clearer.

Reply:
Yea I also agree with Dave and Thomas. I think type synonyms (when used to improve readability) are most useful when they hide the "noise" of a type signature. I don't find the type `Json -> Either String a` to be a noisy. All of that info is vital to understanding the functions inputs and outputs (And the type signature is relatively short in general)

Reply:
Sorry, don't agree

I imagine there are functions everywhere in this file https://github.com/purescript-contrib/purescript-argonaut-codecs/blob/f006470fde11f39d5b17f2f40d05e6af320d6471/src/Data/Argonaut/Decode/Decoders.purs

______

@garyb writes `JsonCodec` everywhere, not `BasicCodec (Either JsonDecodeError) J.Json`

https://github.com/garyb/purescript-codec-argonaut/blob/f8766cb1dcc3c80d712e6c72ce91624dede84038/src/Data/Codec/Argonaut.purs#L57

______

Another pro argument - will be easier to change error type from `String` to typed error, for us and for users

Reply:
@srghma Would you mind making the synonym change in a separate PR so it can be discussed separately? That way I can merge this PR as the other changes are good to go.

Reply:
`BasicCodec` is already 2 levels deep of synonym too :smile:. 

But those codec types aren't on the same level as `Json -> Either String a`, I would be fine with that as-is. I think `F` in `Foreign` is a similar case, and a mistake also, as again the underlying types are pretty simple without the synonym.

Reply:
The example with `F` is good, ok, maybe I'm wrong, will revert

@thomashoneyman no need, I have changed my mind

Reply:
@thomashoneyman done